hedera-node/hedera-app/src/main/java/com/hedera/node/app/grpc/GrpcServiceBuilder.java

```diff
@@ -0,0 +1,177 @@
+ * com.hedera.node.app.workflows.ingest.IngestWorkflow} or {@link
+ * com.hedera.node.app.workflows.query.QueryWorkflow}, so they can do the protobuf parsing. We do
+ * this to segregate the code. This class is <strong>only</strong> responsible for the gRPC call,
+ * the workflows are responsible for working with protobuf.
+ */
+@NotThreadSafe
+public final class GrpcServiceBuilder {
+    /** Logger */
+    private static final Logger LOG = LoggerFactory.getLogger(GrpcServiceBuilder.class);
+
+    /**
+     * Create a single JVM-wide Marshaller instance that simply reads/writes byte arrays to/from
+     * {@link InputStream}s. This class is totally thread safe because it does not reuse byte
+     * arrays. If we get more sophisticated and reuse byte array buffers, we will need to use a
+     * {@link ThreadLocal} to make sure we have a unique byte array buffer for each request.
+     */
+    private static final NoopMarshaller NOOP_MARSHALLER = new NoopMarshaller();
+
+    /**
+     * Create a single instance of the marshaller supplier to provide to every gRPC method
+     * registered with the system. We only need the one, and it always returns the same
+     * NoopMarshaller instance. This is fine to use with multiple app instances within the same JVM.
+     */
+    private static final MarshallerSupplier MARSHALLER_SUPPLIER =
+            new MarshallerSupplier() {
+                @Override
+                public <T> MethodDescriptor.Marshaller<T> get(Class<T> clazz) {
+                    //noinspection unchecked
+                    return (MethodDescriptor.Marshaller<T>) NOOP_MARSHALLER;
+                }
+            };
+
+    /** The name of the service we are building. */
+    private final String serviceName;
+
+    /** The {@link IngestWorkflow} to invoke for transaction methods. */
+    private final IngestWorkflow ingestWorkflow;
+
+    /** The {@link QueryWorkflow} to invoke for query methods. */
+    private final QueryWorkflow queryWorkflow;
+
+    /**
+     * The set of transaction method names that need corresponding service method definitions
+     * generated.
+     */
+    private final Set<String> txMethodNames = new HashSet<>();
+
+    /**
+     * The set of query method names that need corresponding service method definitions generated.
+     */
+    private final Set<String> queryMethodNames = new HashSet<>();
+
+    /**
+     * Creates a new builder.
+     *
+     * @param serviceName The name of the service. Cannot be null or blank.
+     * @param ingestWorkflow The workflow to use for handling all transaction ingestion API calls
+     * @param queryWorkflow The workflow to use for handling all queries
+     */
+    public GrpcServiceBuilder(
+            @Nonnull String serviceName,
+            @Nonnull IngestWorkflow ingestWorkflow,
+            @Nonnull QueryWorkflow queryWorkflow) {
+        this.ingestWorkflow = Objects.requireNonNull(ingestWorkflow);
```

Comment:
I like to do such runtime null checks next to the compiletime annotation checks. Why should define this as best practice since I've seen several places in the code that does not do such checks